docs/api_design.md

```diff
@@ -0,0 +1,69 @@
+# API Design
+
+> Warning: This is work in progress and APIs are very likely to change.
+
+This is a working draft of the API design for **s2-api**.
+
+## Anatomy
+
+```
+s2_api
+|-- Bot
+    |-- Action
+    |-- `tick()`
+|-- Connection
+|-- Engine
+    |-- GameState
+    |-- `run()`
+|-- Remote
+    |-- Controller
+        |-- `act()`, `observe()`, `ping()`, ...
+    |-- Protocol
+        |-- `send()` and `receive()`
+|-- SC2API
+    |-- Request/Response
+        |-- (protocol types)
+```
+
+## Bot
+
+The agent used to infer **GameState** and take an **Action**. The **Bot** will
+be notified of **GameState** via a `tick()`.
+
+### Action
+
+In-game actions that can be performed by a **Bot**.
+
+## Connection
+
+Responsible for establishing the WebSocket connection to a StarCraft II running
+process for sending and receiving messages. Used by the **Remote::Protocol**.
+
+## Engine
+
+Coordinates the connection and change of **GameState** to any **Bot** and
+accepts an **Action** to be sent to the **Remote::Controller** via `run()`.
+
+### GameState
+
+A representation of the current game state of a running StarCraft II process.
+
+## Remote
+
+Handles the request and response cycle for a requested **Bot::Action**
+between the **Engine** and **Connection**.
+
+### Controller
+
+Primarily used to `act()` and `observe()`. Also will be able to make
+`ping()` and `game_info()` types of one-off requests.
+
+### Protocol
+
+Will send and receive **SC2API** request and response via a **Connection**.
```

Comment:
Hmm, this sounds an awful lot like what the `Connection` will be doing. I reckon we can skip the `Protocol` step and the controller contains a `Connection` internally and does things like `self.conn.ping()` and `self.conn.send_action(some_action)`. Then the `Connection` constructs the corresponding protobuf messages internally, converting high level `Action`s into whatever, then it sends that to the websocket client over a channel.

Reply:
It's not clear to me yet who is going to be responsible for the construction the protocol buffers. My initial though was that the`Connection` would have two public functions `send_request(request: raw_protobuf_api::Request)` and `resc_response() -> Result<raw_protobuf_api::Reponse>` and `Protocol` would be the intermediate layer between lib and WebSocket connection. The `Connection` is then only responsible for encoding/decoding messages from a WebSocket.